### ISSUE TITLE
Feature Request: use ollama to redo/improve the OCR

### ISSUE THREAD

**Issue by thiswillbeyourgithub** [icereed/paperless-gpt#20]
Hi,

I've been playing around with using ollama to generate OCR content and was wondering if you were planning on adding the ability to use vision LLM as OCR content for paperless.

I am following several projects that allow using LLM for paperless but none so far use vision LLMs. I recently made a post in an issue for doclytics [here](https://github.com/B-urb/doclytics/issues/97#issuecomment-2425097239) where I outline briefly what's currently possible.

As paperless-gpt seems to have the workflow that best fit my setup, I was wondering what you think of this. That's really the only missing aspect for me to take the jump!

Have a nice day.

**Comment by icereed** [comment]
Hi @thiswillbeyourgithub, I’m very interested to give it a try. I image a third checkbox in the UI saying “Enhance OCR” or similar. What do you think?

**Comment by thiswillbeyourgithub** [comment]
Glad you're interested!

Not bad idea. So the butons would be "refresh" "generate tags" and "enhance OCR"? Maybe sometime it would make sense to run both at the same time, so a fifthbutton "both" would make sense. Or rather instead of buttons use tick boxes and a single button "Execute"?

Btw an update just drop that might motivate you even further :) https://github.com/B-urb/doclytics/issues/97#issuecomment-2425299864

**Comment by icereed** [comment]
Nice! I guess, the first step is to get the images for each page from the paperless API. I don't know if there is a document-to-image API already or if we have to convert PDF to PNG or similar.

**Comment by icereed** [comment]
Converting PDF to images in go doesn't seem to be complicated. Additionally, go is very fast:
https://github.com/Mindinventory/Golang-PDF-to-Image-Converter

**Comment by icereed** [comment]
I just checked and indeed, my paperless is showing the edits made by paperless-gpt in the document history. That's neat!

Currently, the logs are pushed to stdout. With log files come a few challenges regarding log rotation and not spamming the disk of the users. But I do see the point of having a way to backtrack changes.

**Comment by icereed** [comment]
@thiswillbeyourgithub , if you are playful with go you can test around with an early prototype :)

https://github.com/icereed/paperless-gpt/pull/29/files#r1817924426

**Comment by thiswillbeyourgithub** [comment]
Hey thanks a bunch for giving this a try. Unfortunately the only thing I know about go is its name so I won't be of any help at this stage :/

I'm still keeping an eye on this project to know when I can give it a spin and I can't wait :) !

**Comment by icereed** [comment]
@thiswillbeyourgithub now you can try  an early prototype:

It can be used with docker tag icereed/paperless-gpt:unreleased.

**Comment by icereed** [comment]
Ah, sorry, the docker push did not work. The image is building now and should be available in a few minutes. Sorry 😅

**Comment by icereed** [comment]
Yes, you can mix. I personally use ollama for vision and openai for the other suggestions.

**Comment by icereed** [comment]
So usually, when VISION_LLM_PROVIDER and VISION_LLM_MODEL values are set, the UI will show this button:

<img width="1055" alt="image" src="https://github.com/user-attachments/assets/4ce58e85-ac47-45cd-9b94-32bad0ab8e33">

The OCR screen looks like this:

<img width="789" alt="image" src="https://github.com/user-attachments/assets/ae660d79-3702-4cd3-a8f9-917a5b5e946f">

You need to enter a document ID.

The result will look like this:

<img width="766" alt="image" src="https://github.com/user-attachments/assets/fe694791-dbc1-412b-b46e-96d5cb846125">

**Comment by icereed** [comment]
P.S: @mikekaldig be sure to run `docker compose pull` first, the image is hot and fresh :)

**Comment by mikekaldig** [comment]
okay .) fixed for me. But there is a error msg: 
![Screenshot 2024-10-28 213830_222](https://github.com/user-attachments/assets/eec8d076-f65d-4091-9986-74f60a010e65)

**Comment by mikekaldig** [comment]
http://10.10.10.122:8777/documents/325/details

**Comment by icereed** [comment]
This might only be a UI issue, sorry. Can you still click on the button? It should work.

**Comment by icereed** [comment]
P.S: You need to pull the specified ollama model first using `ollama pull x/llama…`

**Comment by mikekaldig** [comment]
Yeah, its only a ui bug. i can click the button:

Great work :)

![Screenshot 1](https://github.com/user-attachments/assets/8fda005b-ef36-47d8-9cf8-5cf0de5df866)

**Comment by icereed** [comment]
Kudos to @thiswillbeyourgithub for bringing in this idea. 💡 
The next step is to integrate the OCR feature into a nice workflow. The OCR takes quite a time on my machine (60s per page minimum). 
In the backend I already designed OCR as an asynchronous job/queue architecture. Maybe that’s something we can put into the background…

**Comment by mikekaldig** [comment]
I just ran a page with completed forms through Ollama 3.2 Vision. The result is noticeably better than standard OCR, especially with the handwritten sections, which are almost perfectly recognized. I think the prompting needs a bit of fine-tuning, but the result is fantastic.

**Comment by JonasHess** [comment]
Nice work!

**Comment by icereed** [comment]
> I just ran a page with completed forms through Ollama 3.2 Vision. The result is noticeably better than standard OCR, especially with the handwritten sections, which are almost perfectly recognized. I think the prompting needs a bit of fine-tuning, but the result is fantastic.

@mikekaldig how fast/slow was the OCR for you and what hardware do you use to run Ollama?

**Comment by JonasHess** [comment]
With my Nvidia Quadro P400  (only 2GB VRAM) it took about 3-4 minutes per page at 300 DPI.

**Comment by icereed** [comment]
> okay .) fixed for me. But there is a error msg: ![Screenshot 2024-10-28 213830_222](https://private-user-images.githubusercontent.com/42111533/380885137-eec8d076-f65d-4091-9986-74f60a010e65.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxODI4ODMsIm5iZiI6MTczMDE4MjU4MywicGF0aCI6Ii80MjExMTUzMy8zODA4ODUxMzctZWVjOGQwNzYtZjY1ZC00MDkxLTk5ODYtNzRmNjBhMDEwZTY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI5VDA2MTYyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVkYzEzMjA5NmMwOTdmZDdkYjZjZjhjMmJjMWU0NzZhOGQ1NjM4MDRmMGFhMDYxYWQxY2I5OTJlNWRkNThjNjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.E2lq6Rm9Lp_8Bm_VnQGOI5Pd6db9IXr0_UG0yogkCEM)

This bug is fixed now with #37 🙂

**Comment by thiswillbeyourgithub** [comment]
Great work it seems! Unfortunately the timing is as bad as can be for me so I don't wait for me to make a review :)

Have you figured out a way to periodically rerun the OCR as better models and prompts are rolled out? For example say I scan my entire collection with the model minicpm, how easy is it to cleanly re OCR my entire collection with llama 3.2 when it's out of beta?

**Comment by mikekaldig** [comment]
> > I just ran a page with completed forms through Ollama 3.2 Vision. The result is noticeably better than standard OCR, especially with the handwritten sections, which are almost perfectly recognized. I think the prompting needs a bit of fine-tuning, but the result is fantastic.
> 
> @mikekaldig how fast/slow was the OCR for you and what hardware do you use to run Ollama?

Hey, i use a Nvidia P40 - Ollama 0.40r5 with Ollama 3.2 Vision, it needs ~ 60 sec for one page

`Every 2.0s: nvidia-smi                                                                                                                                                                                      server1: Tue Oct 29 19:31:17 2024

Tue Oct 29 19:31:17 2024
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.161.08             Driver Version: 535.161.08   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla P40                      Off | 00000000:01:00.0 Off |                  Off |
| N/A   52C    P0              53W / 250W |  11350MiB / 24576MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A   3424810      C   ...unners/cuda_v12/ollama_llama_server    11348MiB |
+---------------------------------------------------------------------------------------+
WARNING: infoROM is corrupted at gpu 0000:01:00.0

`

**Comment by mikekaldig** [comment]
I think every user probably has a different way of handling this. Personally, I prefer it to be processed sequentially in the background. But maybe you could alternatively implement a switch function that allows for personalization. I also think you're probably now getting into the topic of whether there should be a settings area. :)

**Comment by icereed** [comment]
Another thing I wonder: some documents have perfect contents already since they come in as digital documents (PDFs). We could theoretically let an LLM decide if it’s worth it to do an OCR again. But that could also be too much unexpected magic for the users…

**Comment by thiswillbeyourgithub** [comment]
Personally, my use case would be to run the entire collection into it sequentially in the background. and run the whole thing all over again when there is a new model

**Comment by JonasHess** [comment]
> Personally, my use case would be to run the entire collection into it sequentially in the background. and run the whole thing all over again when there is a new model

This is my use case as well. I also don't really have a need for the web ui.

**Comment by icereed** [comment]
So basically, a tag would be sufficient then as an MVP. Just tag all documents that you wanna OCR and paperless-gpt will put them into the job queue.

**Comment by JonasHess** [comment]
> So basically, a tag would be sufficient then as an MVP. Just tag all documents that you wanna OCR and paperless-gpt will put them into the job queue.

It should also add a custom tag when it is done, so another pipeline can be triggered after.

**Comment by icereed** [comment]
Interesting, tag chaining 😄
I will give it a thought.

**Comment by thiswillbeyourgithub** [comment]
> So basically, a tag would be sufficient then as an MVP. Just tag all documents that you wanna OCR and paperless-gpt will put them into the job queue.

I agree ideally customizable because paperless-gpt is quite longe and can clutter some paperless UI I'd say

**Comment by thiswillbeyourgithub** [comment]
Also how is handled the fact that some documents have some fields manually edited already? Does it overwrite them? Ignore them?

**Comment by thiswillbeyourgithub** [comment]
Btw ollama 0.4 just dropped and has support for larger llama3.2 vision models

**Comment by thiswillbeyourgithub** [comment]
Hi, I just came looking for news of this feature

**Comment by icereed** [comment]
Hey @thiswillbeyourgithub,

what would your favorite use case be?
Is it an automatic OCR based on tags?

**Comment by thiswillbeyourgithub** [comment]
Basically a "smarter OCR and classification".

You see I take photos of many unexpected document (medical reports? Cards? Pictures of objects?) and having a vision model able to both do a more sane OCR as well as tag and choose a title seems like it would greatly helo my setup.

**Comment by Timtam** [comment]
Hello everyone, thought I'd chime in with some questions.
I just set up paperless-gpt and find it to be very, very helpful and would love it to further develop into a really great companion to paperless-ngx.
Currently what I'd love to happen here is an automatic OCR mode, like the one we already have for tagging / titling. That'd really help. Plus, a nicer UI that doesn't require me to input a document ID would really be appreciated ;).
In addition to that, does anyone have experience with x/llama3.2-vision? I just ran it once yesterday, and it told me something along the lines of "I skipped transcribing since this section contained a handwritten signature". Especially this part would have been interesting to me, since i'm blind and OCRing handwritten text, although probably errorneous, is particularly important to me. Is there anything we can do about this, like changing the prompt to tell it to still transcribe this part?
Thanks, have a great time, we'll see us in 2025.

**Comment by icereed** [comment]
Hey all :)
I finally had some time to introduce an automatic OCR mode:
https://github.com/icereed/paperless-gpt/pull/75
